xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -83,16 +98,42 @@
   @VisibleForTesting
   static boolean enableTimeout =
       Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_ENABLE_TIMEOUT"));
+  @VisibleForTesting
+  static boolean enableFaultInjection =
+      Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_FAULT_INJECTION"));
+  @VisibleForTesting
+  static final Metadata.Key<String> DOWNSTREAM_NODE_KEY =
+      Metadata.Key.of("x-envoy-downstream-service-node", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_HTTP_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_GRPC_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-grpc-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  private static final AtomicLong activeFaultInjectedStreamCounter = new AtomicLong();
 
   private final XdsLogger logger;
   private final String authority;
   private final ServiceConfigParser serviceConfigParser;
   private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService scheduler;
   private final XdsClientPoolFactory xdsClientPoolFactory;
   private final ThreadSafeRandom random;
   private final ConcurrentMap<String, AtomicInteger> clusterRefs = new ConcurrentHashMap<>();
   private final ConfigSelector configSelector = new ConfigSelector();
 
+  @VisibleForTesting
+  AtomicLong activeFaultInjectedStreams = activeFaultInjectedStreamCounter;
```

Comment:
Why not make `activeFaultInjectedStreamCounter` mutable and create a setter for test (similar to [DnsNameResolver](https://github.com/grpc/grpc-java/blob/3c4d3d2bb2303d041594202b3ccac85aa15352c7/core/src/main/java/io/grpc/internal/DnsNameResolver.java#L129))? Having two fields is odd.

Reply:
If parallel testing is a concern, inject on non-static counter. Injecting on static counter does not work.

If parallel testing is not a concern, just test with static final counter, no need to inject.

Reply:
> If parallel testing is a concern, inject on non-static counter. Injecting on static counter does not work.

This definitely is a concern. Each test case should never have side effect that could potentially affect the result of other test cases. Otherwise, other test cases may observe weird behavior and it would be extremely hard to find out the cause.

I was saying you could have a setter that overrides the static atomic in the test and reset it at the end of the test (same as what you do for the static flag). Still this isn't a good approach. Using a static factory for the atomic and inject it into the resolver solves the problem (as I mentioned for `SharedCallCounterMap`). Anyway, if you found the test is not testable, that means the implementation has problems.

Btw, I don't see where you test the behavior of max_active_faults.

Reply:
> I was saying you could have a setter that overrides the static atomic in the test and reset it at the end of the test (same as what you do for the static flag)

This approach does not solve the parallel testing problem.

> if you found the test is not testable, that means the implementation has problems

Injecting static is not testable if there is parallel testing. Injecting private counter is simple and testable.

Reply:
> Btw, I don't see where you test the behavior of max_active_faults.

`resolved_faultDelayWithMaxActiveStreamsInLdsUpdate()` is the test.